### ISSUE TITLE
pymalloc is not aware of Memory Tagging Extension (MTE) and crashes

### ISSUE THREAD

**Issue by cc37f5d2-8e01-467f-856d-61531acd8e9a** [python/cpython#87759]
BPO | [43593](https://bugs.python.org/issue43593)
--- | :---
Nosy | @tim-one, @nascheme, @vstinner, @tiran

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-03-22.13:29:47.118>
labels = ['library', '3.9']
title = 'pymalloc is not aware of Memory Tagging Extension (MTE) and crashes'
updated_at = <Date 2021-04-04.18:17:33.327>
user = 'https://bugs.python.org/ggardet'
```

bugs.python.org fields:
```python
activity = <Date 2021-04-04.18:17:33.327>
actor = 'tim.peters'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2021-03-22.13:29:47.118>
creator = 'ggardet'
dependencies = []
files = []
hgrepos = []
issue_num = 43593
keywords = []
message_count = 13.0
messages = ['389312', '389313', '389315', '389316', '389318', '389686', '389693', '389783', '389928', '389943', '390135', '390143', '390206']
nosy_count = 5.0
nosy_names = ['tim.peters', 'nascheme', 'vstinner', 'christian.heimes', 'ggardet']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue43593'
versions = ['Python 3.9']
```

</p></details>

### Linked PRs
* gh-113737

**Comment by diegorusso** [comment]
Hello everyone, this issue picked my interest and decided to investigate what's going on.

TLDR;
The PR https://github.com/python/cpython/pull/14474 from @vstinner fixes indeed the issue with MTE. It's only Python 3.9 (and earlier versions) to be affected as the PR landed to 3.10. Here you can find the back port to 3.9.
This has been validate on a software emulation platform with ubuntu 23.04 and MTE enabled. MTE is a security feature hence the commit can be considered as security fix.

Longer version:
In order to reproduce the bug I had to use a "recent" version of Ubuntu (23.04) as 22.04 has glibc with MTE support but it has been disabled by default.
I've tested all the latest tags of Python (3.9 to 3.12 and main) and only 3.9 is failing when MTE is enabled (`export GLIBC_TUNABLES=glibc.mem.tagging=3`)

With Python 3.9.18 you have this behaviour:
```
~/cpython-no-radix-tree$ export GLIBC_TUNABLES=glibc.mem.tagging=3
~/cpython-no-radix-tree$ ./python
Segmentation fault
```
Next thing is to see what's going on with gdb. I know the `address_in_range` causes issue hence I put a breakpoint in there.
```
~/cpython-no-radix-tree$ export GLIBC_TUNABLES=glibc.mem.tagging=3
~/cpython-no-radix-tree$ gdb -ex "b address_in_range" -ex "run" ./python
GNU gdb (Ubuntu 13.1-2ubuntu2.1) 13.1
...
Breakpoint 1, address_in_range (p=p@entry=0xfffff78df330, pool=pool@entry=0xfffff78df000) at Objects/obmalloc.c:1420
c
1420        uint arenaindex = *((volatile uint *)&pool->arenaindex);
(gdb) c
Continuing.

Breakpoint 1, address_in_range (p=p@entry=0xfffff78df330, pool=pool@entry=0xfffff78df000) at Objects/obmalloc.c:1420
1420        uint arenaindex = *((volatile uint *)&pool->arenaindex);
(gdb)
Continuing.

Breakpoint 1, address_in_range (p=p@entry=0xfffff78df330, pool=pool@entry=0xfffff78df000) at Objects/obmalloc.c:1420
1420        uint arenaindex = *((volatile uint *)&pool->arenaindex);
(gdb)
Continuing.

Breakpoint 1, address_in_range (p=p@entry=0xfffff78df330, pool=pool@entry=0xfffff78df000) at Objects/obmalloc.c:1420
1420        uint arenaindex = *((volatile uint *)&pool->arenaindex);
(gdb)
Continuing.

Breakpoint 1, address_in_range (p=p@entry=0xfffff78e1030, pool=pool@entry=0xfffff78e1000) at Objects/obmalloc.c:1420
1420        uint arenaindex = *((volatile uint *)&pool->arenaindex);
(gdb)
Continuing.

Breakpoint 1, address_in_range (p=p@entry=0x600fffff7c65a20, pool=pool@entry=0x600fffff7c65000) at Objects/obmalloc.c:1420
1420        uint arenaindex = *((volatile uint *)&pool->arenaindex);
(gdb)
Continuing.

Program received signal SIGSEGV, Segmentation fault
Memory tag violation while accessing address 0x0600fffff7c65020
Allocation tag 0x0
Logical tag 0x6.
address_in_range (p=p@entry=0x600fffff7c65a20, pool=pool@entry=0x600fffff7c65000) at Objects/obmalloc.c:1420
1420        uint arenaindex = *((volatile uint *)&pool->arenaindex);
(gdb)
Continuing.

Program terminated with signal SIGSEGV, Segmentation fault.
The program no longer exists.
(gdb) r
Starting program: /home/user/cpython-no-radix-tree/python
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/aarch64-linux-gnu/libthread_db.so.1".

Breakpoint 1, address_in_range (p=p@entry=0xfffff78df330, pool=pool@entry=0xfffff78df000) at Objects/obmalloc.c:1420
1420        uint arenaindex = *((volatile uint *)&pool->arenaindex);
(gdb)
(gdb)
(gdb) c
Continuing.

Breakpoint 1, address_in_range (p=p@entry=0xfffff78df330, pool=pool@entry=0xfffff78df000) at Objects/obmalloc.c:1420
1420        uint arenaindex = *((volatile uint *)&pool->arenaindex);
(gdb)
Continuing.

Breakpoint 1, address_in_range (p=p@entry=0xfffff78df330, pool=pool@entry=0xfffff78df000) at Objects/obmalloc.c:1420
1420        uint arenaindex = *((volatile uint *)&pool->arenaindex);
(gdb)
Continuing.

Breakpoint 1, address_in_range (p=p@entry=0xfffff78df330, pool=pool@entry=0xfffff78df000) at Objects/obmalloc.c:1420
1420        uint arenaindex = *((volatile uint *)&pool->arenaindex);
(gdb)
Continuing.

Breakpoint 1, address_in_range (p=p@entry=0xfffff78e1030, pool=pool@entry=0xfffff78e1000) at Objects/obmalloc.c:1420
1420        uint arenaindex = *((volatile uint *)&pool->arenaindex);
(gdb)
Continuing.

Breakpoint 1, address_in_range (p=p@entry=0x700fffff7c65a20, pool=pool@entry=0x700fffff7c65000) at Objects/obmalloc.c:1420
1420        uint arenaindex = *((volatile uint *)&pool->arenaindex);
(gdb)
Continuing.

Program received signal SIGSEGV, Segmentation fault
Memory tag violation while accessing address 0x0700fffff7c65020
Allocation tag 0x0
Logical tag 0x7.
address_in_range (p=p@entry=0x700fffff7c65a20, pool=pool@entry=0x700fffff7c65000) at Objects/obmalloc.c:1420
1420        uint arenaindex = *((volatile uint *)&pool->arenaindex);
(gdb) bt
#0  address_in_range (p=p@entry=0x700fffff7c65a20, pool=pool@entry=0x700fffff7c65000) at Objects/obmalloc.c:1420
#1  0x0000aaaaaab49a28 in pymalloc_free (ctx=<optimized out>, p=p@entry=0x700fffff7c65a20) at Objects/obmalloc.c:1866
#2  0x0000aaaaaab4b2a8 in _PyObject_Free (ctx=<optimized out>, p=0x700fffff7c65a20) at Objects/obmalloc.c:1920
#3  0x0000aaaaaab4a3f4 in _PyMem_DebugRawFree (ctx=ctx@entry=0xaaaaaae9cd60 <_PyMem_Debug+96>, p=p@entry=0x700fffff7c65a30) at Objects/obmalloc.c:2212
#4  0x0000aaaaaab4a434 in _PyMem_DebugFree (ctx=0xaaaaaae9cd60 <_PyMem_Debug+96>, ptr=0x700fffff7c65a30) at Objects/obmalloc.c:2345
#5  0x0000aaaaaab4b600 in PyObject_Free (ptr=ptr@entry=0x700fffff7c65a30) at Objects/obmalloc.c:709
#6  0x0000aaaaaab34f98 in dictresize (mp=mp@entry=0xfffff78df0b0, minsize=<optimized out>) at Objects/dictobject.c:1285
#7  0x0000aaaaaab34fb0 in insertion_resize (mp=mp@entry=0xfffff78df0b0) at Objects/dictobject.c:1040
#8  0x0000aaaaaab3bef8 in PyDict_SetDefault (d=0xfffff78df0b0, key=key@entry=0xfffff78de3b0, defaultobj=defaultobj@entry=0xfffff78de3b0) at Objects/dictobject.c:2964
#9  0x0000aaaaaaba6b10 in PyUnicode_InternInPlace (p=p@entry=0xffffffffeb10) at Objects/unicodeobject.c:15726
#10 0x0000aaaaaaba6eb4 in PyUnicode_InternFromString (cp=<optimized out>) at Objects/unicodeobject.c:15762
#11 0x0000aaaaaab6bdbc in _PyTypes_InitSlotDefs () at Objects/typeobject.c:7600
#12 0x0000aaaaaab45620 in _PyTypes_Init () at Objects/object.c:1736
#13 0x0000aaaaaabf8b54 in pycore_init_types (tstate=tstate@entry=0xf00fffff7c655c0) at Python/pylifecycle.c:588
#14 0x0000aaaaaabf943c in pycore_interp_init (tstate=0xf00fffff7c655c0) at Python/pylifecycle.c:725
#15 0x0000aaaaaabf95c0 in pyinit_config (runtime=runtime@entry=0xaaaaaaee01a0 <_PyRuntime>, tstate_p=tstate_p@entry=0xffffffffef50, config=config@entry=0xffffffffed70)
    at Python/pylifecycle.c:766
#16 0x0000aaaaaabfadb8 in pyinit_core (runtime=runtime@entry=0xaaaaaaee01a0 <_PyRuntime>, src_config=src_config@entry=0xfffffffff000, tstate_p=tstate_p@entry=0xffffffffef50)
    at Python/pylifecycle.c:927
#17 0x0000aaaaaabfaf3c in Py_InitializeFromConfig (config=config@entry=0xfffffffff000) at Python/pylifecycle.c:1134
#18 0x0000aaaaaaafd940 in pymain_init (args=args@entry=0xfffffffff208) at Modules/main.c:66
#19 0x0000aaaaaaafda44 in pymain_main (args=args@entry=0xfffffffff208) at Modules/main.c:698
#20 0x0000aaaaaaafdb3c in Py_BytesMain (argc=<optimized out>, argv=<optimized out>) at Modules/main.c:731
#21 0x0000aaaaaaafc660 in main (argc=<optimized out>, argv=<optimized out>) at ./Programs/python.c:15
(gdb)
```

Looking at the GDB output we see that all the addresses that are passed to `address_in_range` starts with `0xfffff`. At some point a couple of addresses are passed with the tag: in the first execution the address starts with 0x600fffff whist in the second execution starts with 0x700fffff. 6 and 7 are the MTE tags that the malloc returned.

It always fails after six times `address_in_range` is called: this because python is trying to increase the size of a dictionary
The dictresize increases the size of the dictionary and it goes over the size of the arena and it revert to the malloc to allocate memory. the malloc returns an address with the tag because MTE is enable. So far so good.

The issue is when we are going to check the address with the tag with `address_in_range` but very likely if the object was not allocated by `pymalloc` then the `poolp` pointer we get by rounding down the address to the nearest 512 byte boundary will have the wrong tag and trigger an MTE violation when we deference it.

With the radix tree implementation, we never dereference the pointer passed to address_in_range or any pointer derived from it so no MTE violation can occur.

I strongly suggest to back port the radix tree implementation to 3.9 as it will be supported until 2025. MTE is a security feature (https://community.arm.com/arm-community-blogs/b/architectures-and-processors-blog/posts/enhanced-security-through-mte) on Arm platforms to protect the user from invalid memory accesses.
Not having the radix tree implementation in 3.9 doesn't allow the users to use this security feature. Because of this we could consider this as a security fix.

**Comment by diegorusso** [comment]
Hello, after backporting the implementation of the radix tree to 3.9 branch, I've received a some feedback highlighting that this is not a security fix and not even a bug fix because it is actually a new feature (thanks @pablogsal and @nascheme).

In the last couple of days I've given some extra thoughts to the issue/PR and I've concluded that the best course of action is **not to merge** this (and subsequent fixes) and leave as it is.

This because of the benefits that we have from this PR are very little compared to the cost of integrating it: it is not just merging this PR but (as @nascheme said in the PR) all subsequent fixes.

Even though we do it, a patch release of 3.9 needs to be created and it eventually will be  propagated to distros. When the distros will pick up the update, those will be old LTS where glibc anyway is not MTE ready so almost impossible to use it.

glibc support for MTE has been introduced in 2.32. Ubuntu 22.04 has glibc 2.35 but MTE is disabled by default. We need to update to 23.04 with glibc 2.37 to have MTE enabled by default in glibc. In ubuntu 23.04 the default Python version is 3.11.

It is very unlikely to have a case where a fairly recent distribution is used with an older version of Python on it. At that point if the user really needs MTE, we can simply suggest to use a newer version of Python.

If we really want to do anything here, we could just update the 3.9 documentation by acknowledging that Python 3.9 is not compatible with MTE on aarch64 and the workaround is to use a more updated version of Python.
The simplest option is not even to update the documentation and leave the user ending up in this issue and find out to use a newer version of Python :)

Thoughts?

**Comment by diegorusso** [comment]
I suggest closing the issue as it has been decided not to progress with the backport to 3.9 anymore.
